### ISSUE TITLE
Ungraceful system shutdown

### ISSUE THREAD

**Issue by simplysoft** [dappnode/DAppNode#379]
When shutting down or rebooting dappnode, blockchain state can be lost (e.g. by geth) because processes are not properly / gracefully stopped. 

**To Reproduce**

1. Trigger reboot command (via dappnode UI, ssh or via ACPI)
2. Shutdown happens almost instantly (feels way to quick compared to other systems)
3. After Startup, e.g. check geth logs: 
```
WARN [XX-XX|XX:XX:XXXXX] Old unclean shutdowns found              count=XX
WARN [XX-XX|XX:XX:XXXXX] Unclean shutdown detected                booted=XX age=XXX
```

**Expected behavior**

All docker containers should be stopped before system proceeds with shutdown. 
Stopping geth before shutdown via docker stop does not show the same problem, so the system must shutdown without stopping containers or proceeding with shutdown before they stopped  

**DAppNode version:**

Core DAppNode Packages versions
bind.dnp.dappnode.eth: 0.2.6
dappmanager.dnp.dappnode.eth: 0.2.45, commit: 7b9ad9ad
https.dnp.dappnode.eth: 0.1.2
ipfs.dnp.dappnode.eth: 0.2.15
wifi.dnp.dappnode.eth: 0.2.8
wireguard.dnp.dappnode.eth: 0.1.1

System info
dockerComposeVersion: 1.25.5
dockerServerVersion: 20.10.6
dockerCliVersion: 20.10.6
os: debian
versionCodename: bullseye
architecture: amd64
kernel: 5.10.0-10-amd64
Disk usage: 62%

**Additional context**
Installed via DAppNodeISO

**Comment by dapplion** [comment]
@simplysoft Thanks for opening the issue! DAppNode should give Geth 20 minutes of grace time to persist in-memory data and properly shutdown https://github.com/dappnode/DAppNodePackage-geth/blob/37925e10cce3379ad30e5202c592efdf4f14cfd5/dappnode_package.json#L12

From your issue that is not the case, so @3alpha can you investigate please? Try recreating locally if you can

**Comment by CMTRACE** [comment]
I've seen this for an entire year or more. On discord it was explained to me months ago it was expected behaviour.

Glad to see it's not actually it never felt right.

**Comment by 3alpha** [comment]
This issue arises because there is difference between `docker stop` timeout and timeout when docker daemon is stopped. It should be fixed by adding line
```
{
...
"shutdown-timeout": 600,
...
}
```
to the `/etc/docker/daemon.json` config file. I'll test it and report back.

**Comment by pablomendezroyo** [comment]
If we are about to edit the timeout of the docker daemon, we should warn users in the UI that it may take that long for shutting down dappnode, otherwise it will be an unexpected behavior.

**Comment by Pol-Lanski** [comment]
> If we are about to edit the timeout of the docker daemon, we should warn users in the UI that it may take that long for shutting down dappnode, otherwise it will be an unexpected behavior.

I agree, if the user actually presses the button on their machine and it takes 20 minutes to shut down, they will be definitely not expecting that. Maybe a solution is to leave physical shutdown (via power button) in a shorter time-frame and do the software shutdown (via UI) as long as it needs to be - warning the user that it might take time?

**Comment by CMTRACE** [comment]
I’d be expecting the system to shut down gracefully even with a power button press. If I wanted to ungracefully shut a machine down, I’d yank or press and hold the power button like any other PC in the last 15 years.



________________________________

**Comment by 3alpha** [comment]
600 seconds was just an example, it takes significantly shorter amount of time. I agree with @CMTRACE.

**Comment by CMTRACE** [comment]
Just bumping this issue, there is open conversation happening on the stakers channel about this with quite a few people looking to use a ups with dappnode. 

The simplest way I can picture this working is if any shutdown command, button press, or ui interaction gracefully shuts down the containers. 

This way a supported UPS or NUT server can just issue a the shutdown command to the os and it waits patiently for the containers to close. 

Once power is restored the nuc firmware will power on the system once again when mains is detected

**Comment by 3alpha** [comment]
I have been working on this issue a lot, but I haven't been able to find satisfactory solution. Problem is that docker ignores timeout flags when daemon itself is shutting down.

**Comment by alexpeterson91** [comment]
> I have been working on this issue a lot, but I haven't been able to find satisfactory solution. Problem is that docker ignores timeout flags when daemon itself is shutting down. In all other cases stopping is done gracefully.

So having the daemon shutdown ungracefully is still causing and issue?  I see how there's some things it could affect. But if the containers are just all stopped gracefully, aren't most issues gonna be mitigated already?  And if the daemon needs to restart when power is determined to be back on or the outage was temporary/a big surge and power is back on or stable again. Could you just have the startup command for the containers be preceded by a `systemctl docker restart` command to start with a fresh daemon and then restart the previously running containers?
